### DISCUSSION TITLE
Can one authorize REST API client to specific realm only

### DISCUSSION THREAD

**Discussion by christian-2** [keycloak/keycloak#22347] [Q&A] [answered]
The Keycloak [Server Developer Guide](https://www.keycloak.org/docs/latest/server_development/#admin-rest-api) explains how one can authenticate to the Admin REST API either 

1. with an OAuth 2.0 [Resource Owner Password Credentials Grant](https://datatracker.ietf.org/doc/html/rfc6749#section-4.3) and username/password for user `admin` in realm `master` or
2. with an OAuth 2.0 [Client Credentials Grant](https://datatracker.ietf.org/doc/html/rfc6749#section-4.4) and a client (again in realm `master`) that holds certain permissions (e.g. has the role `admin` assigned to it).

Is it possible to work alternatively with less expansive permissions (below the level of `admin`)? For instance, I would like to restrict a REST client to operations in a specific realm (other than `master`) only.

**Comment by CollinEstes** [comment]
I'm not a dev on the project but a long time user. You can provision a client on your non master realm and go with the client credentials grant approach where you give that client the roles specific to only that realm that you would like it to have. The examples are for the master realm but it hold true for any other realm as well.

**Comment by christian-2** [comment]
@CollinEstes Thanks for your response; I've now tried the following. It leads me to believe that the procedure only almost works like for realm `master`, but not quite. Whether this is a feature or a bug, I currently don't know.

```
url=https://... # Keycloak 21.1.2
realm=my-realm # added role admin, as in realm master
client_id=my-client # client in that realm; added audience mapper and service account role, as prescribed
client_secret=... # client secret of that client
access_token=$(curl -s -X POST "$url/realms/$realm/protocol/openid-connect/token" \
  -d "grant_type=client_credentials&client_id=$client_id&client_secret=$client_secret" | jq -r .access_token) 
curl -v -H "Authorization: bearer $access_token" "$url/admin/realms/$realm" # 200, as expected
curl -v -H "Authorization: bearer $access_token" "$url/admin/realms/$realm/clients" # 200, as expected
curl -v -H "Authorization: bearer $access_token" "$url/admin/realms/$realm/authentication/flows" # 200, as expected
curl -v -H "Authorization: bearer $access_token" "$url/admin/realms/$realm/users" # 403, unexpected (this would work for master)
curl -v -H "Authorization: bearer $access_token" "$url/admin/realms/master" # 403, as expected
```

**Comment by danielFesenmeyer** [comment]
You have (at least) two ways to authenticate at a target realm (different from master):

- login as a user from master realm. In this case, you use `master` realm urls for login. You can restrict permissions to the target-realm, by just requesting roles from the special client `<target-realm-name>-realm`.
- login as a user from the target realm. In this case, you use `<target-realm-name>` realm urls for login. Permissions are automatically restricted to the target-realm. You can further restrict permissions by requesting only specific roles from the target-realm's _realm-management_ client (which are mostly the same as the roles of master realm's client `<target-realm-name>-realm`).

**Comment by christian-2** [comment, marked as answer]
This change in the Keycloak Administration Console solved my issue: `my-realm` / `my-client` / Service accounts roles / Assign role / Filter by clients: `realm-management view-users`.
```
curl -v -H "Authorization: bearer $access_token" "$url/admin/realms/$realm/users" # 200, as expected
curl -v -H "Authorization: bearer $access_token" "$url/admin/realms/master/users" # 403, as expected
```

Notice that realm role `default-roles-my-realm` was already assigned to `my-client` by default and that it includes e.g. `realm-managment view-clients`. Also notice that changed role assignments apparently only become effective for newly issued access tokens.
